File: plugins/probe-rat-layer/probe_layer/io/wire_gen.py
"""飞线生成：2-pin 单线 / 3-pin share / ≥4-pin MST。

供 JSON loader 与 xlsx loader 共用。
"""
from __future__ import annotations

import math

from ..model import Net, Pin, Wire, NetClass

_SHARE_SHORT_IGNORE = 10.0   # 3-pin share 的短段（圆内）短于此值则忽略


def _wire(n: Net, i: int, a: Pin, b: Pin) -> Wire:
    return Wire(f"{n.net_id}_W{i}", n.net_id, a.pos, b.pos, n.width, n.clearance)


def share_wires(n: Net, start_i: int) -> list[Wire]:
    """3-pin share：外 pin → 较远内 pin 长段 + 两内 pin 间短段。

    直角 L 形（外→内拐点→内）：长段按扇区分层，短段在圆内、太短可忽略。
    """
    outer = max(n.pins, key=lambda p: math.hypot(p.pos.x, p.pos.y))
    inners = [p for p in n.pins if p is not outer]
    far = max(inners, key=lambda p: p.pos.dist(outer.pos))
    other = inners[0] if inners[0] is not far else inners[1]
    wires = [_wire(n, start_i, outer, far)]
    if far.pos.dist(other.pos) >= _SHARE_SHORT_IGNORE:
        wires.append(_wire(n, start_i + 1, far, other))
    return wires


def mst_wires(n: Net, start_i: int) -> list[Wire]:
    """多 pin net（≥4）用 Prim 最小生成树连成飞线，总长更短、交叉更少。

    O(n²)，n 为 pin 数（通常 ≤9），无需空间索引。
    """
    pins = n.pins
    npins = len(pins)
    in_tree = [False] * npins
    dist = [math.inf] * npins
    parent = [-1] * npins
    dist[0] = 0.0
    wires: list[Wire] = []
    for _ in range(npins):
        u = min((i for i in range(npins) if not in_tree[i]), key=lambda i: dist[i])
        in_tree[u] = True
        if parent[u] != -1:
            wires.append(_wire(n, start_i + len(wires), pins[parent[u]], pins[u]))
        for v in range(npins):
            if not in_tree[v]:
                d = pins[u].pos.dist(pins[v].pos)
                if d < dist[v]:
                    dist[v] = d
                    parent[v] = u
    return wires


def generate_wires(nets: tuple[Net, ...], warnings: list[str]) -> tuple[Wire, ...]:
    """无显式 wires 时生成飞线：2-pin 单线 / 3-pin share / ≥4-pin MST。

    只给 signal net 生成；power/ground 走 plane 层。
    """
    wires: list[Wire] = []
    for n in nets:
        if n.net_class != NetClass.SIGNAL:   # 电源/地走 plane 层，不生成飞线
            continue
        if len(n.pins) < 2:
            warnings.append(f"网络 {n.net_id} 仅有 {len(n.pins)} 个 pin，跳过")
            continue
        i0 = len(wires)
        if len(n.pins) == 2:
            wires.append(_wire(n, i0, n.pins[0], n.pins[1]))
        elif len(n.pins) == 3:
            wires.extend(share_wires(n, i0))
        else:
            wires.extend(mst_wires(n, i0))
    return tuple(wires)

File: core-plugins/probe-rat-layer/src/io/wire_gen.rs
//! 飞线生成：2-pin 单线 / 3-pin share / ≥4-pin MST（移植自 Python `io/wire_gen.py`）。

use crate::model::{Net, NetClass, Pin, Wire};
use std::f64::INFINITY;

const _SHARE_SHORT_IGNORE: f64 = 10.0;

fn _wire(n: &Net, i: usize, a: &Pin, b: &Pin) -> Wire {
    Wire::new(
        format!("{n}_W{i}", n = n.net_id),
        n.net_id.clone(),
        a.pos,
        b.pos,
        n.width,
        n.clearance,
    )
}

/// 3-pin share：外 pin → 较远内 pin 长段 + 两内 pin 间短段。
pub fn share_wires(n: &Net, start_i: usize) -> Vec<Wire> {
    let outer = n
        .pins
        .iter()
        .max_by(|p, q| {
            (p.pos.x.hypot(p.pos.y)).partial_cmp(&(q.pos.x.hypot(q.pos.y))).unwrap()
        })
        .unwrap();
    let inners: Vec<&Pin> = n.pins.iter().filter(|p| p.pin_id != outer.pin_id).collect();
    let far = inners
        .iter()
        .max_by(|p, q| p.pos.dist(outer.pos).partial_cmp(&q.pos.dist(outer.pos)).unwrap())
        .unwrap();
    let other = if inners[0].pin_id != far.pin_id {
        inners[0]
    } else {
        inners[1]
    };
    let mut wires = vec![_wire(n, start_i, outer, far)];
    if far.pos.dist(other.pos) >= _SHARE_SHORT_IGNORE {
        wires.push(_wire(n, start_i + 1, far, other));
    }
    wires
}

/// 多 pin net（≥4）用 Prim 最小生成树连成飞线。
pub fn mst_wires(n: &Net, start_i: usize) -> Vec<Wire> {
    let pins = &n.pins;
    let npins = pins.len();
    let mut in_tree = vec![false; npins];
    let mut dist = vec![INFINITY; npins];
    let mut parent = vec![-1i64; npins];
    dist[0] = 0.0;
    let mut wires: Vec<Wire> = Vec::new();
    for _ in 0..npins {
        let mut u = None;
        let mut best = INFINITY;
        for i in 0..npins {
            if !in_tree[i] && dist[i] < best {
                best = dist[i];
                u = Some(i);
            }
        }
        let u = u.unwrap();
        in_tree[u] = true;
        if parent[u] != -1 {
            wires.push(_wire(n, start_i + wires.len(), &pins[parent[u] as usize], &pins[u]));
        }
        for v in 0..npins {
            if !in_tree[v] {
                let d = pins[u].pos.dist(pins[v].pos);
                if d < dist[v] {
                    dist[v] = d;
                    parent[v] = u as i64;
                }
            }
        }
    }
    wires
}

/// 无显式 wires 时生成飞线：2-pin 单线 / 3-pin share / ≥4-pin MST。
/// 只给 signal/power net 生成（电源走信号层一起分层）；ground 已在加载时剔除。
pub fn generate_wires(nets: &[Net], warnings: &mut Vec<String>) -> Vec<Wire> {
    let mut wires: Vec<Wire> = Vec::new();
    for n in nets {
        if !matches!(n.net_class, NetClass::Signal | NetClass::Power) {
            continue;
        }
        if n.pins.len() < 2 {
            warnings.push(format!("网络 {} 仅有 {} 个 pin，跳过", n.net_id, n.pins.len()));
            continue;
        }
        let i0 = wires.len();
        match n.pins.len() {
            2 => wires.push(_wire(n, i0, &n.pins[0], &n.pins[1])),
            3 => wires.extend(share_wires(n, i0)),
            _ => wires.extend(mst_wires(n, i0)),
        }
    }
    wires
}
